### DISCUSSION TITLE
Upgrade moosefs3 to moosefs4

### DISCUSSION THREAD

**Discussion by seycob** [moosefs/moosefs#626] [Q&A]
I have a moosefs debian12 cluster, goal 3

1 master, 2 metalogger, 10 chunkservers and 3 clients. 3.0.118

And 3 more clients on 3.0.117 , these last ones are 32 bit systems and I can't change them at the moment .....

My question is if I can upgrade the rest of the machines to version 4 leaving the 32 bit clients on 3.0.117 and they still access correctly.

And if the correct procedure would be, stop the cluster, backup configurations and metadata, change the apt repository to moosefs4, perform the upgrade.

I will not use EC for the moment, the new configuration files of the different machines must be reconfigured or the old ones are compatible ? I access the CGI from apache.

Any indication will be welcome, Thank you
![mfsclient](https://github.com/user-attachments/assets/94966fd5-c1c0-40c0-9519-a3247d0dfb75)

**Comment by anon314159** [comment]
You can have multiple versions of different clients connected to the master but any attempts to enforce or create new storage classes or goals that leverage Erasure Codes will fail. The default settings allow for interoperability between different versions.  Furthermore, I would consider upgrading your clients to support 64bit microprocessors as there's a very good chance major distributions and the mainline kernel will eventually drop support in the not so distant future.  This has a domino effect nearly every other project compiled for 32bit systems.

**Comment by seycob** [comment]
Thanks for the clarification, all the machines involved except for those 3 clients are x86_64.
The reason to keep those three mfsclient is only operational, backup programs on certain older OS, on remote users, fail to send data to vsftpd on x86_64, all suspicions fall on the encryption method on connections and certificates, current.
I have tested several times on x86_64 and at least 200 users do not progress, they have OS, windows7, windows10 32 bit, .... And their backup programs are old x86_32.
I try to find solutions for those customers and keep the clusters up to date. I may have to send them to another type of storage ......

**Comment by anon314159** [comment]
Assuming you are not bandwidth/throughput limited why not export the chunkservers as SMB/NFS shares for your legacy Windows clients?  NFS-ganesha works very well with MooseFS and so long as you have an AD/LDAP environment mapping UIDs/GIDs (ACLs) shouldn't pose a major hurdle.

MFS Export -> FUSE Mount (SMB/NFS Proxy Server) -> SMBD/NFS-ganesha (VFS) -> Windows Client.

**Comment by chogata** [comment]
@seycob You can use old clients with an upgraded MooseFS instance. As long as you keep to the rule that master is in the same or newer version than chunkservers, and master and chunkservers are in the same or newer versions than clients, the system will work.

We do not recommend to keep different versions of master and chunkservers for long, it should only be an upgrade situation (but if the whole upgrade procedure takes a couple of days, that's not a major problem).

Clients are much more resilient and generally older clients have no problem working with a much newer master. Of course, a newer client will probably be better and faster, but if you are bound by an older OS, they can still be used.

You can use EC on your instance, as long as all the files that will be accessed by version 3 clients are NOT kept in that format. For example: you can have a storage class in copies only and files needed by version 3 clients in that class, and another class with EC format, but with files not accessible to mountpoints used by version 3 clients.

An attempt to access or modify an EC file with version 3 client will result in errors on the client side and may lead to unexpected behaviours (it will not alter the file). So making sure that users cannot access such files and cannot change storage classes to classes with EC is recommended.

Currently the only way to ensure that a user cannot change a storage class is to take away the possibility at all (from that one export, you can have several exports defined and different clients using different export definitions to mount moosefs).

We have a new version almost ready to release, that has some new features, including the possibility to limit classes that users can use in an export, so that will be something that will help with this kind of configuration.

**Reply by seycob** [reply]
All masters, chunkservers, metalogers and 3 mfsclients are on version 3.0.118 x86_64 and I have no limitation, remote users can not change anything and I do not need EC, I can increase chunkservers or deploy more clusters, are the three remaining mfclients on which vsftpd runs and where certain remote backup clients upload data, for which I have not found a consistent solution for now because it depends on remote backup clients.
As far as I understand I can upgrade to version 4 and leave the remaining mfsclients on 3.0.117 for some time.

**Comment by struthio** [comment]
I had mix of v3 and v4 installations. Everything worked fine until I defined first EC storage_class.
After I've created first EC in master, 'v3 chunkservers and clients' refused to connect to master, so then I had to update all of them to v4.

**Reply by anon314159** [reply]
> I had mix of v3 and v4 installations. Everything worked fine until I defined first EC storage_class. After I've created first EC in master, 'v3 chunkservers and clients' refused to connect to master, so then I had to update all of them to v4.

Please reference this link for the underlying cause and solution. 

https://github.com/moosefs/moosefs/issues/609

**Comment by seycob** [comment]
Iniciare el upgrade a la version 4 siguiendo los consejos recibidos.
